File: matrix/49wordSearch/wordSearch.java
public class wordSearch {
    static boolean backTrack(char[][] board, String word, int i, int j, int ROWS, int COLS, int idx, int length) {
        if (idx >= length) {
            return true;
        }
        if (i < 0 || j < 0 || i >= ROWS || j >= COLS || board[i][j] != word.charAt(idx)) {
            return false;
        }
        boolean exist = false;
        if (board[i][j] == word.charAt(idx)) {
            board[i][j] += 100;
            exist = (backTrack(board, word, i + 1, j, ROWS, COLS, idx + 1, length) ||
                    backTrack(board, word, i, j + 1, ROWS, COLS, idx + 1, length) ||
                    backTrack(board, word, i - 1, j, ROWS, COLS, idx + 1, length) ||
                    backTrack(board, word, i, j - 1, ROWS, COLS, idx + 1, length));
            board[i][j] -= 100;
        }
        return exist;
    }

    static boolean search(char[][] board, String word) {
        int ROWS = board.length, COLS = board[0].length, length = word.length();
        for (int i = 0; i < ROWS; i++) {
            for (int j = 0; j < COLS; j++) {
                if (backTrack(board, word, i, j, ROWS, COLS, 0, length)) {
                    return true;
                }
            }
        }
        return false;
    }

    public static void main(String[] args) {
        char[][] board = {
                { 'A', 'B', 'C', 'E' },
                { 'S', 'F', 'C', 'S' },
                { 'A', 'D', 'E', 'E' }
        };
        String word = "ABCB";
        System.out.println(search(board, word));
    }
}

File: matrix/49wordSearch/wordSearch.py
class Solution:
    def search(self, board: list[list[str]], word: str) -> bool:
        ROWS, COLS = len(board), len(board[0])
        path = set()
        def backTrack(r, c, idx):
            if idx == len(word):
                return True
            if (r < 0 or c < 0 or r >= ROWS or c >= COLS or word[idx] != board[r][c] or (r,c) in path):
                return False
            path.add((r,c))
            res = (backTrack(r + 1, c, idx + 1) or
                   backTrack(r - 1, c, idx + 1) or
                   backTrack(r, c + 1, idx + 1) or
                   backTrack(r, c - 1, idx + 1))
            path.remove((r,c))
            return res
        for r in range(ROWS):
            for c in range(COLS):
                if backTrack(r, c, 0):
                    return True
        return False
print(Solution().search([["A", "Z", "O", "L"], ["N", "E", "V", "P"], ["W", "Q", "D", "X"]], "LOVE"))